### PR TITLE
Adds a germline segment genotyper tool.

### PR DESCRIPTION
**Pull request by vruano**
This new tool takes the segments discovered by the discovery tool and ask the question whether the same event is present in all the input samples.
The results ia a VCF quite similar to the one produced by XHMM.

Closes issue #147.

### CODE REVIEW

**Comment by vruano** [comment]
@davidbenjamin please take a look at your earliest convenience.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/CopyNumberTriStateSegmentCaller.java [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+     * Cached value of ln(10)^-1
+     */
+    protected static final double INV_LN_10 = 1.0 / Math.log(10);
+
+    /**
+     * Maximum reportable output quality score; higher quality scores will
+     * capped to this value.
+     */
+    public static final double MAX_QUAL_SCORE = 199.99999999999;
+
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/CopyNumberTriStateSegmentCaller.java]: extra blank line.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/CopyNumberTriStateSegmentCaller.java [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+                standardizeBySample(counts);
+                break;
+            case TARGET:
+                standardizeByTarget(counts);
+                break;
+        }
+    }
+
+    private final void standardizeByTarget(final RealMatrix counts) {
+        final double[] rowMeans = IntStream.range(0, counts.getRowDimension())
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/CopyNumberTriStateSegmentCaller.java]: Need to use the GATKProtectedMathUtils.columnMeans, StdDev, rowMeans, StDev methods. This changed were lost when rebasing.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/DiscoverCopyNumberTriStateSegments.java [hunk 3]
```diff
@@ -119,51 +71,10 @@ protected Object doWork() {
...
-            case SAMPLE:
-                standardizeBySample(counts);
-                break;
-            case TARGET:
-                standardizeByTarget(counts);
-                break;
-        }
-    }
-
-    private void standardizeByTarget(final RealMatrix counts) {
```
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/DiscoverCopyNumberTriStateSegments.java]: Here is actual code, so just need to copy it over.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/CopyNumberTriStateSegmentCaller.java [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+
+    @Override
+    protected final Object doWork() {
+        final CopyNumberTriStateHiddenMarkovModel model = modelArguments.createModel();
+        final TargetCollection<Target> targets = targetArguments.readTargetCollection(false);
+        final ReadCountCollection inputCounts = readAndSortCountsByTargetCoordinates(targets);
+        applyZScoreTransformation(inputCounts);
+        checkForMissingTargetsInInputCounts(targets, inputCounts);
+        openOutput(outputFile, model, targets, inputCounts);
+        final Object result = makeCalls(model, targets, inputCounts);
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/CopyNumberTriStateSegmentCaller.java]: `makeCalls` always returns `SUCCESS` as long as it doesn't throw an exception, right?  So why not make `makeCalls` a `void` method and let `doWork` return `SUCCESS`?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/GenotypeCopyNumberTriStateSegments.java [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+                              final TargetCollection<Target> targets, ReadCountCollection inputCounts) {
+        outputWriter = GATKVariantContextUtils.createVCFWriter(outputFile, null, false);
+        outputWriter.writeHeader(composeHeader(inputCounts.columnNames()));
+    }
+
+    private VCFHeader composeHeader(final List<String> sampleNames) {
+        final VCFHeader result = new VCFHeader(Collections.emptySet(), sampleNames);
+
+        result.addMetaDataLine(new VCFHeaderLine(VCFHeaderVersion.VCF4_2.getFormatString(), VCFHeaderVersion.VCF4_2.getVersionString()));
+        result.addMetaDataLine(CNVAllele.DEL.altHeaderLine("Represents a deletion respect to the expected number of copies"));
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/GenotypeCopyNumberTriStateSegments.java]: *with* respect to.

Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/GenotypeCopyNumberTriStateSegments.java]: I would prefer "reference copy number" over "expected number of copies."

File: src/main/java/org/broadinstitute/hellbender/tools/exome/GenotypeCopyNumberTriStateSegments.java [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+    @Override
+    protected void closeOutput(final File outputFile) {
+        outputWriter.close();
+    }
+
+    @Override
+    protected Object makeCalls(final CopyNumberTriStateHiddenMarkovModel model, final TargetCollection<Target> targets, final ReadCountCollection inputCounts) {
+        logger.info("Composing list of segment intervals to genotype ...");
+        final List<GenotypingSegment> segments = composeGenotypingSegments(segmentsFile, targets);
+        logger.info(String.format("A total of %d segments to genotyping found", segments.size()));
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/GenotypeCopyNumberTriStateSegments.java]: to genotype

File: src/main/java/org/broadinstitute/hellbender/tools/exome/GenotypeCopyNumberTriStateSegments.java [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+                        .map(a -> a.allele).collect(Collectors.toList()));
+
+        CNVAllele(final CopyNumberTriState state, final boolean isRef) {
+            this.state = state;
+            allele = isRef ? Allele.create("N", true) : Allele.create("<" + name() + ">");
+        }
+
+        public VCFSimpleHeaderLine altHeaderLine(final String description) {
+            if (allele.isReference()) {
+                throw new UnsupportedOperationException("you cannot ask for a ALT header line on a reference allele");
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/GenotypeCopyNumberTriStateSegments.java]: an ALT header

**Comment by davidbenjamin** [comment]
@vruano Looks good, merge after fixing a few typos.

**Comment by vruano** [comment]
Squashed, will merge once test pass again.

**Pull request merged**